File: modules/common/src/main/java/com/media/bus/common/entity/common/BaseEntity.java
package com.media.bus.common.entity.common;

import jakarta.persistence.Column;
import jakarta.persistence.Id;
import jakarta.persistence.MappedSuperclass;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;
import org.hibernate.proxy.HibernateProxy;

import java.util.Objects;
import java.util.UUID;

/// 모든 엔티티의 식별자와 동등성(equals/hashCode)을 관리하는 최상위 공통 엔티티입니다.
/// 하이버네이트 프록시를 안전하게 처리하도록 지원합니다.
@Getter
@SuperBuilder
@NoArgsConstructor
@AllArgsConstructor
@MappedSuperclass
public abstract class BaseEntity {

    @Id
    @UuidV7
    @Column(name = "id", updatable = false, nullable = false)
    private UUID id;

    @Override
    public final boolean equals(Object o) {
        if (this == o)
            return true;
        if (o == null)
            return false;
        Class<?> oEffectiveClass = o instanceof HibernateProxy
            ? ((HibernateProxy) o).getHibernateLazyInitializer().getPersistentClass()
            : o.getClass();
        Class<?> thisEffectiveClass = this instanceof HibernateProxy
            ? ((HibernateProxy) this).getHibernateLazyInitializer().getPersistentClass()
            : this.getClass();
        if (thisEffectiveClass != oEffectiveClass)
            return false;
        BaseEntity that = (BaseEntity) o;
        return getId() != null && Objects.equals(getId(), that.getId());
    }

    @Override
    public final int hashCode() {
        return this instanceof HibernateProxy
            ? ((HibernateProxy) this).getHibernateLazyInitializer().getPersistentClass().hashCode()
            : getClass().hashCode();
    }
}

File: modules/common/src/main/kotlin/com/media/bus/common/entity/common/BaseEntity.kt
package com.media.bus.common.entity.common

import org.jetbrains.exposed.v1.core.dao.id.EntityID
import org.jetbrains.exposed.v1.dao.java.UUIDEntity
import java.util.*

/**
 * ## UUID PK 기반 DAO 엔티티 공통 추상 클래스
 */
abstract class BaseEntity(id: EntityID<UUID>) : UUIDEntity(id)

/**
 * ## 감사 컬럼 포함 DAO 엔티티 공통 추상 클래스
 *
 * - `table` 파라미터: 이 엔티티가 매핑되는 `DateBaseTable` 구현체
 * - `var updatedAt`은 서비스 계층에서 변경 시 직접 갱신한다
 */
abstract class DateBaseEntity(id: EntityID<UUID>, table: DateBaseTable) : BaseEntity(id) {
    val createdAt by table.createdAt
    var updatedAt by table.updatedAt
}
